phoenix-core/src/main/java/org/apache/phoenix/expression/function/CollationKeyFunction.java

```diff
@@ -0,0 +1,233 @@
+ * 
+ * It uses the open-source grammaticus and i18n packages to obtain the collators
+ * it needs.
+ * 
+ * @author snakhoda
+ *
+ */
+@FunctionParseNode.BuiltInFunction(name = CollationKeyFunction.NAME, args = {
+		// input string
+		@FunctionParseNode.Argument(allowedTypes = { PVarchar.class }),
+		// ISO Code for Locale
+		@FunctionParseNode.Argument(allowedTypes = { PVarchar.class }, isConstant = true),
+		// whether to use special upper case collator
+		@FunctionParseNode.Argument(allowedTypes = { PBoolean.class }, defaultValue = "false", isConstant = true),
+		// collator strength
+		@FunctionParseNode.Argument(allowedTypes = { PInteger.class }, defaultValue = "null", isConstant = true),
+		// collator decomposition
+		@FunctionParseNode.Argument(allowedTypes = { PInteger.class }, defaultValue = "null", isConstant = true) })
+public class CollationKeyFunction extends ScalarFunction {
+
+	private static final Log LOG = LogFactory.getLog(CollationKeyFunction.class);
+
+	public static final String NAME = "COLLKEY";
+
+	public CollationKeyFunction() {
+	}
+
+	public CollationKeyFunction(List<Expression> children) throws SQLException {
+		super(children);
+	}
+
+	@Override
+	public boolean evaluate(Tuple tuple, ImmutableBytesWritable ptr) {
+		try {
+			String inputValue = getInputValue(tuple, ptr);
+			String localeISOCode = getLocaleISOCode(tuple, ptr);
+			Boolean useSpecialUpperCaseCollator = getUseSpecialUpperCaseCollator(tuple, ptr);
+			Integer collatorStrength = getCollatorStrength(tuple, ptr);
+			Integer collatorDecomposition = getCollatorDecomposition(tuple, ptr);
+
+			Locale locale = LocaleUtils.get().getLocaleByIsoCode(localeISOCode);
+			
+			if(LOG.isDebugEnabled()) {
+				LOG.debug(String.format("Locale: " + locale.toLanguageTag()));
+			}
+			
+			LinguisticSort linguisticSort = LinguisticSort.get(locale);
+
+			Collator collator = BooleanUtils.isTrue(useSpecialUpperCaseCollator)
+					? linguisticSort.getUpperCaseCollator(false) : linguisticSort.getCollator();
+
+			if (collatorStrength != null) {
+				collator.setStrength(collatorStrength);
+			}
+
+			if (collatorDecomposition != null) {
+				collator.setDecomposition(collatorDecomposition);
+			}
+
+			if(LOG.isDebugEnabled()) {
+				LOG.debug(String.format("Collator: [strength: %d, decomposition: %d], Special-Upper-Case: %s",
+					collator.getStrength(), collator.getDecomposition(), BooleanUtils.isTrue(useSpecialUpperCaseCollator)));
+			}
+			
+			byte[] collationKeyByteArray = collator.getCollationKey(inputValue).toByteArray();
+
+			if(LOG.isDebugEnabled()) {
+				LOG.debug("Collation key bytes:" + Arrays.toString(collationKeyByteArray));
+			}
+			
+			// byte is signed in Java, but we need unsigned values for comparison
+			// https://www.programcreek.com/java-api-examples/index.php?api=java.text.CollationKey
+			// Byte.toUnsignedInt will convert a byte value between [-128,127] to an int value
+			// between [0,255]
```

Comment:
This code should not be necessary. Phoenix sorting is based on lexicographical order of the bytes. So -128 is really like 0 and 127 is like 255.

Reply:
I started out with a straight byte array (VARBINARY), but realized the issue is related to 2s complement binary representation.

Here's what I was able to gather:

The collator's collation key produces a byte array that holds values that are supposed to be treated as unsigned. However, since byte is a signed (2s complement) datatype in Java, doing a Java comparison on them with regular operators (which is presumably what Phoenix does) produces incorrect results.

For example, 00000001 represents 1 and 11111111 represents -1 in 2s complement. If we simply compare those bytes using standard operators then 11111111 is less than 00000001. But a collator key byte array that returns these two bytes would like them to be treated as "1" and "255" respectively and would need 11111111 to be greater than 00000001.

Getting this effect out of a byte array basically means taking each byte and widening it (to an int) which results in the lower order bytes representing the unsigned positive number that was intended by the collation key.

See this example usage of toByteArray, which does the same thing: http://icu-project.org/apiref/icu4j59m1/com/ibm/icu/text/CollationKey.html#toByteArray--

BTW, Byte.toUnsignedInt in Java 8 is basically a bitwise AND with 0xFF.

Reply:
> doing a Java comparison on them with regular operators (which is presumably what Phoenix does) produces incorrect results

That's not what Phoenix does. That's what I tried to explain in my previous comment. In your unit tests, the best way to confirm this is through an end-to-end integration test that uses an ORDER BY COLLKEY(my_varchar_col):
* Create a simple table CREATE TABLE t (k INTEGER PRIMARY KEY, v VARCHAR)
* Populate the table with some multi-byte string values:
    UPSERT INTO t VALUES(1,'ണഫɰɸ')
    UPSERT INTO t VALUES(2,'ɚɦ')
* Execute an ORDER BY on the table: SELECT * FROM t ORDER BY COLLKEY(v)
* Confirm that the rows are ordered as expected by a linguistic sort

The above test is a must-have. We can also have lower level unit tests that don't need to run as an integration test to test the sorting. You'd need to manually create an OrderedResultIterator. See OrderedResultIteratorTest as an example.

Reply:
Yes, I'm working on the tests you mentioned and your bulleted list is the sequence I had in mind.

My comment above was trying to explain that without the conversion of the collator key byte array to an array of unsigned integers the ORDER BY does not work as expected. My comment on Phoenix behaviour was speculative, without having looked inside. Introducing conversion to unsigned resulted in the behaviour we need.

Reply:
Let's start with the tests. That conversion should not be necessary.